io/io/src/TKey.cxx

```diff
@@ -339,6 +338,10 @@ TKey::TKey(const void *obj, const TClass *cl, const char *name, Int_t bufsize, T
    lbuf       = fBufferRef->Length();
    fObjlen    = lbuf - fKeylen;
 
+   // Append to mother directory only after the call to Streamer() was
+   // successful (and didn't throw).
+   fCycle = fMotherDir->AppendKey(this);
```

Comment:
This change looks good.   As a side note, `fCycle` is used in the 'write key itself' step on old line 332 (and thus switch from writing correct value to writing default value), however this first streaming is 'just' to reserve the space and it is written again (which updated values for fCycle and the sizes) right before returning out of this function.   Additional side note: I did not review thoroughly whenever the other functions between here and the user class Streamer are exception safe but a cursory look says that they are likely to be.